Standalone Jenkinsfile

michaelsembwever opened the pull request:
CASSANDRA-18594

michaelsembwever committed:
Standalone Jenkinsfile

 – ensure test file paths, and their suite names are unique (the latter was broken for cqlshlib and python dtests)
 – removal of properties and system-out in test xml happens in CassandraXMLJUnitResultFormatter
 – new jenkins agent labels and introduce agents sizes
 - ci_summary.html generation script, ref work submitted in apache/cassandra-builds#99
 – fix for stress-test and fqltool-test running on small
 - ant generate-test-report is limited to only running on individual test types (ci_parser.py provides ci_summary.html for the overview now)
 - each cell has a single retry, and the retry will happen on a different agent
 - on ci-cassandra the summary stage happens on the builtin, bc copyArtifacts on 15k+ files takes many hours otherwise
 - test-burn only needs two splits
 - dependency-check is disabled from the lint target until CASSANDRA-19213
 - add $DEBUG env var to in-tree scripts, turns on bash debug
 - fix FBUtilities' handling of gcp cos_containerd (kernel version comes with a trailing '+' character)

 patch by Aleks Volochnev, Mick Semb Wever; reviewed by Aleksandr Volochnev, Josh McKenzie, Maxim Muzafarov, Stefan Miklosovic for CASSANDRA-18594

Co-authored-by:  Aleksandr Volochnev <a.volochnev@gmail.com>
Co-authored-by:  Mick Semb Wever <mck@apache.org>
Co-authored-by: Josh McKenzie <jmckenzie@apache.org>

michaelsembwever commented:
@HadesArchitect

michaelsembwever commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,480 @@
+    string(name: 'branch', defaultValue: env.BRANCH_NAME, description: 'Branch')
+    choice(name: 'architecture', choices: archsSupported + "all", description: 'Pick architecture. The ARM64 is disabled by default at the moment.')
+    choice(name: 'jdk', choices: jdksSupported + "all", description: 'Pick JDK versions.')
+    booleanParam(name: 'stage_artifacts', defaultValue: false, description: 'Disable to exlude stage') // return default to true
+    booleanParam(name: 'stage_lint', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_debian', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_redhat', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_fqltool-test', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_cqlsh-test', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test-cdc', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test-trie', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test-compression', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_stress-test', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test-burn', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_long-test', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test-oa', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_test-system-keyspace-directory', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_jvm-dtest', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_jvm-dtest-upgrade', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_simulator-dtest', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-novnode', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-offheap', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-large', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-large-novnode', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade-novnode', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade-large', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade-novnode-large', defaultValue: false) // return default to true
```
this still needs a fix:
- all post-commit stages must be enabled by default (so post-commit builds just run without prompting)
- simpler UX (clicking a gazillion checkboxes is no good)

michaelsembwever replied:
post-commit is always hard to test in advance.

I'm counting on this:

> When that job is triggered, and is configured with parameters, it does have default parameters.

https://stackoverflow.com/questions/26017620/how-to-run-parametrized-build-after-commit-to-repository

michaelsembwever commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,480 @@
+    checkout changelog: false, scm: scmGit(branches: [[name: params.branch]], extensions: [cloneOption(depth: 1, noTags: true, reference: '', shallow: true)], userRemoteConfigs: [[url: params.repository]])
+    copyArtifacts filter: 'test/**/TEST*.xml.xz', fingerprintArtifacts: true, projectName: env.JOB_NAME, selector: specific(env.BUILD_NUMBER), target: "build/", optional: true
+    if (fileExists('build/test/output')) {
+      sh ".build/docker/_docker_run.sh bullseye-build.docker generate-test-report.sh"
```
needs a fix
- TESTS-TestSuites.xml is empty
- TESTS-html.tar is empty

michaelsembwever replied:
fixed.

michaelsembwever commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,480 @@
+    stage('Summary') {
+      steps {
+        script {
+          if(stageResults.find { "FAILURE" == it.value }) {
+            currentBuild.result='FAILURE'
+            error("Build failed due to failed stages")
+          } else if(stageResults.find { "UNSTABLE" == it.value || "ABORTED" == it.value }) {
+            if ("ABORTED" != currentBuild.result) { currentBuild.result='UNSTABLE' }
+            echo("Build unstable due to unstable or aborted stages")
```
on a failed/aborted build we're getting 
```
MissingPropertyException: No such property: stageResults for class: groovy.lang.Binding
```

but I'm not sure where it's coming from… 
see https://ci-cassandra.apache.org/job/Cassandra-5-devbranch/59/console

michaelsembwever commented on src/java/org/apache/cassandra/utils/FBUtilities.java:
```diff
@@ -1405,15 +1406,21 @@ static Semver getKernelVersionFromUname()
         if (!isLinux)
             return null;
```
it's kinda odd that we can return null here, but we fail parsing gcp cos_containerd's uname

michaelsembwever commented on src/java/org/apache/cassandra/utils/FBUtilities.java:
```diff
@@ -1429,13 +1436,19 @@ static Semver getKernelVersionFromUname()
+
+                if (version.endsWith("+"))
+                        // gcp's cos_containerd has a trailing +
+                        version = StringUtils.chop(version);
```
new approach.  gcp's cos_containerd is now supported. 
fyi @jacek-lewandowski

smiklosovic commented on .build/check-code.sh:
```diff
@@ -24,5 +24,5 @@ command -v ant >/dev/null 2>&1 || { echo >&2 "ant needs to be installed"; exit 1
 
 # execute
-ant -f "${CASSANDRA_DIR}/build.xml" check dependency-check
+ant -f "${CASSANDRA_DIR}/build.xml" check # dependency-check # FIXME dependency-check now requires NVD key downloaded first
```
we could probably workaround it somewhere when this gets merged.

https://github.com/apache/cassandra/pull/3191/files

We might stop to require key if we just point it to a mirror.

HadesArchitect replied:
@michaelsembwever Is it urgent/important? I could investigate.

Mmuzaf replied:
This is related to the following issue:
https://issues.apache.org/jira/browse/CASSANDRA-19146

The NVD database should be cached somewhere on the agent's side (~94,6 MB on disk on my local machine), however, I'm not sure that it is an appropriate thing for Jenkins agents.

smiklosovic commented on .build/ci/requirements.txt:
```diff
@@ -0,0 +1,18 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+beautifulsoup4
```
Recently, we started to explicitly set versions for requirements.txt for python dtests. Would it be appropriate to set versions explicitly here as well? That way we are sure that as new versions of these libraries are released, it will not be automagically updated and then it might start to make problems when some changes in newer versions are not compatible anymore.

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,518 @@
 // Validate/lint this file using the following command
 // `curl -X POST  -F "jenkinsfile=<.jenkins/Jenkinsfile" https://ci-cassandra.apache.org/pipeline-model-converter/validate`
 
+
```
Should we use 4 spaces for a tab or 2 spaces? The code varies from one style to another, I think it should be consistent (4 tabs?).

michaelsembwever replied:
yeah, good comment @Mmuzaf

michaelsembwever replied:
commits pushed. please check.

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,518 @@
+//// scripting support ////
+///////////////////////////
+
+def jdksSupported() { return ["11", "17"] }
```
I don't think this should be hardcoded, but fetched from the `build.xml`. We already have the same in properties there:
```
    <property name="java.default" value="11" />
    <property name="java.supported" value="11,17" />
```

We can use the ant cli tool to print it or create a dedicated goal for it e.g. `ant print-property`

michaelsembwever replied:
We wanted to take this approach, but this function needs to work outside of the context of a a current (git cloned) workspace. 

What you describe is provided by `javaVersionsSupported()` (and `javaVersionsSupported()`), but we have forgotten to use them as intended: the build should fail if the hardcoded values don't match those from build.xml

michaelsembwever replied:
> the build should fail if the hardcoded values don't match those from build.xml

given recent changes
- turning tasks() into a method instead of global variables, and
- changing the 'jdk' parameter to be "what jdks to restrict the build to", rather that "what jdks to build with"

it's easy now to completely remove the hardcoded methods, but it means changing the 'jdk' parameter from choice to string.

michaelsembwever replied:
commits pushed. please check.

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,518 @@
+        it.value.put('splits', 1)
+    }
+
+    def testSteps = [
```
I think we should add a `print` command to `run-tests.sh` (or `docker/run-tests.sh`) and populate this list from the output of that command.

michaelsembwever replied:
yes.

that will happen as part of [CASSANDRA-18731](https://issues.apache.org/jira/browse/CASSANDRA-18731).
(but it won't be run-tests.sh that is responsible for defining the list)

Mmuzaf commented on .build/run-tests.sh:
```diff
@@ -240,6 +241,9 @@ _main() {
       fi
       ant testclasslist -Dtest.classlistprefix=distributed -Dtest.timeout=$(_timeout_for "test.distributed.timeout") -Dtest.classlistfile=<(echo "${testlist}") ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
       ;;
+    "build_dtest_jars")
```
Should we add `dtest-latest` to the `run-tests.sh` as well?

michaelsembwever replied:
It is:
- https://github.com/apache/cassandra/blob/19509dbd623af9fdbf1696694e41034b9b6b40f6/.build/run-python-dtests.sh#L120-L121
- https://github.com/apache/cassandra/blob/19509dbd623af9fdbf1696694e41034b9b6b40f6/.build/docker/run-tests.sh#L147

Mmuzaf replied:
I meant this one `.build/run-tests.sh`, but now I see that it's not designed for the dtests, or should it be? The `"jvm-dtest-upgrade" | "jvm-dtest-upgrade-novnode"` is there at least.

Mmuzaf commented on .build/run-python-dtests.sh:
```diff
@@ -164,10 +165,13 @@ fi
 cat ${TMPDIR}/nosetests.xml > ${DIST_DIR}/test/output/nosetests.xml
-ant -quiet -silent generate-unified-test-report
+sed "s/testsuite name=\"Cassandra dtests\"/testsuite name=\"${DTEST_TARGET}_jdk${java_version}_python${python_version}_cython${cython}_$(uname -m)${SPLIT_STRING}\"/g" ${DIST_DIR}/test/output/nosetests.xml > ${TMPDIR}/nosetests.xml
+cat ${TMPDIR}/nosetests.xml > ${DIST_DIR}/test/output/nosetests.xml
```
Let's declare `${DIST_DIR}/test/output/` as a constant, it's used many times throughout the file

Mmuzaf replied:
I think we can also consider moving it to the `build.xml`

michaelsembwever replied:
not sure exactly what you mean by moving it to `build.xml` (but i'd suggest doing whatever that is as a separate ticket – this patch is blocking the 5.0-beta2 release)

michaelsembwever replied:
pushed, check new commits.

Mmuzaf commented on .build/ci/generate-ci-summary.sh:
```diff
@@ -0,0 +1,49 @@
+...
+EOL
+
+pip install -r ${CASSANDRA_DIR}/.build/ci/requirements.txt
```
This should be a part of the `bullseye-build.docker` image, isn't it? Do we need to run it every time? As far as I can see this script is never called outside of a docker container.

michaelsembwever replied:
pushed, check new commits.

Mmuzaf commented on .build/ci/logging.sh:
```diff
@@ -0,0 +1,124 @@
+export SILENCE_LOGGING=false
+export LOG_TO_FILE="${LOG_TO_FILE:-false}"
+
+disable_logging() {
```
Let's encapsulate it in `precommit_check.sh` where it is used, and remove methods that are not used. For now, we don't need them all. We can add new methods later when we need them.

michaelsembwever replied:
@jmckenzie-dev what do you think ?

Mmuzaf commented on .build/ci/precommit_check.sh:
```diff
@@ -0,0 +1,49 @@
+# limitations under the License.
+
+
+source "logging.sh"
```
Let's inline it:
https://github.com/apache/cassandra/pull/2852/files#r1545303799

Mmuzaf commented on .build/build-checkstyle.xml:
```diff
@@ -41,7 +41,7 @@
             <formatter type="plain"/>
             <formatter type="xml" tofile="${checkstyle.report.file}"/>
-            <fileset dir="${build.src.java}" includes="**/*.java"/>
+            <fileset dir="${build.src.java}" includes="**/*.java" excludes="build/**/*.java" />
```
As far as I understand there shouldn't be the `build` directory under the `src`. Why do we need this change?

michaelsembwever replied:
@HadesArchitect ? 

I will (re-) check this…

michaelsembwever replied:
removing it.

michaelsembwever replied:
pushed, check new commits.

Mmuzaf commented on .build/ci/generate-test-report.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh -e
```
This script is a pure `ant` target, so I think we can move it under the `.build` directory and rename to `run-generate-test-report.sh`.

michaelsembwever replied:
it's only used by CI (it depends on structuring of files that is currently happening in `.jenkins/Jenkinsfile` 
…?

Mmuzaf replied:
We can move the whole `ci' folder under `.jenkins', as nobody uses it except Jenkins itself.

michaelsembwever replied:
they are intended to be standards (and/or helper scripts) for all CI systems.

`.build/ci/generate-test-report.sh` is to aggregate all test xml files.  it is intended to be used all all splits on a given test type/target/

`.build/ci/generate-ci_summary.sh` is to create the ci_summary.html sharable artefact. it expects a folder hierarchy that separates test types/targets.  (it will locally link each failure and testsuite to the junit html file found in this hierarchy, as it was generated by the above script.)  This file and the python files it uses are originally from a non-jenkins CI.

Adding this information in as header comments to each file

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,534 @@
+    it.value.put('splits', 1)
+  }
+
+  def testSteps = [
```
This could have a huge impact on us soon, as we will have to update this list every time a new destination is added in three files: here, `run-tests.sh` and `docker/run-tests.sh`. The complexity is quite high and increases.

Mmuzaf replied:
I don't have a good idea how we can avoid this :-(

michaelsembwever replied:
agree.

this is to be addressed in [CASSANDRA-18731](https://issues.apache.org/jira/browse/CASSANDRA-18731)

Mmuzaf commented on .build/docker/_docker_run.sh:
```diff
@@ -118,11 +122,11 @@ docker_command="export ANT_OPTS=\"-Dbuild.dir=\${DIST_DIR} ${CASSANDRA_DOCKER_AN
 # re-use the host's maven repository
 container_id=$(docker run --name ${container_name} -d --security-opt seccomp=unconfined --rm \
-    -v "${cassandra_dir}":/home/build/cassandra -v ~/.m2/repository/:/home/build/.m2/repository/ -v "${build_dir}":/dist \
+    -v "${cassandra_dir}":/home/build/cassandra -v ${m2_dir}:/home/build/.m2/repository/ -v "${build_dir}":/dist \
```
Inside the `docker/run-tests.sh` we are mapping these params in the following way:
```
docker_mounts="-v 
${cassandra_dir}:/home/cassandra/cassandra -v
"${build_dir}":/home/cassandra/cassandra/build -v 
${m2_dir}:/home/cassandra/.m2/repository"
```

and here the mapping is different, should it be the same?

michaelsembwever replied:
there are intended differences here yes.
(maybe we could look at the /dist approach for the build_dir, but that's out-of-scope for this ticket)

michaelsembwever replied:
the reason for the m2_dir  override in jenkins is maven does not support multiple builds downloading at the same time (believe it or not).
we can't share the same ~/.m2/repository to multiple containers, or writing to the same file leaves corruption.


this (the need for this info) came up here: https://the-asf.slack.com/archives/CK23JSY2K/p1779458154665669?thread_ts=1778491370.688919&cid=CK23JSY2K

Mmuzaf commented:
We should update the `.build/README.md` as we've removed `dtest-offheap` and added `dtest-latest`. It's worth mentioning `build-jars.sh` as well.

michaelsembwever commented:
> We should update the .build/README.md as we've removed dtest-offheap and added dtest-latest.

Doing this.

> It's worth mentioning build-jars.sh as well.

Not doing this (it doesn't add value, it's only used by CI to create the stash)

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,534 @@
+  for (def axis in _axes) {
+    def cell = axis
+    def name = getStepName(cell, stepsMap[cell.step])
+    tasks[cell.step == "jar" ? "jars" : "tests"][name] = { ->
```
Why do we need such a mapping? Can't we use `'jars'` anywhere else in this file?

michaelsembwever replied:
tasks are either "jars" or "tests"

jars is used in the first stage here: https://github.com/apache/cassandra/pull/2852/files/19509dbd623af9fdbf1696694e41034b9b6b40f6#diff-917491029f94ac1ce6c114a542d2a8738ce1f2d1e866eaffe7189d2139250200R72-R74

tests are used in the second stage here: https://github.com/apache/cassandra/pull/2852/files/19509dbd623af9fdbf1696694e41034b9b6b40f6#diff-917491029f94ac1ce6c114a542d2a8738ce1f2d1e866eaffe7189d2139250200R84

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,534 @@
+      // merge splits for each target's test report, other axes are kept separate
+      //   TODO parallelised for loop
+      //   TODO results_details.tar.xz needs to include all logs for failed tests
+      sh """
```
We can inline this script into the `generate-ci-summary.sh` as it is used only for Jenkins.

michaelsembwever replied:
Not really.  There's structuring of files this is dependent on that only happens in this Jenkinsfile.

I do believe there's opportunity to push this along with the path structuring of test result files down.  But this has to be done in consultation with other CI systems (that changes in build.xml and the scripts would impact).

Also note the challenge here with docker-in-docker…

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +11,534 @@
+  }
+}
+
+//  CASSANDRA-18130
```
This will be a part of another issue, so there is no need to keep it in this PR.

michaelsembwever replied:
removed 👍

michaelsembwever commented:
>> We should update the .build/README.md as we've removed dtest-offheap and added dtest-latest.
>
>Doing this.

pushed, check new commits.

michaelsembwever commented on .build/README.md:
```diff
@@ -29,66 +29,6 @@ The following applies to all build scripts.
-    SONAR_PROJECT_KEY="<key of the Cassandra project in SonarQube>"
-
-The analysis can be run with `ant sonar`.
-
```
moved to the bottom, given it's experimental

Mmuzaf commented on .build/ci/ci_parser.py:
```diff
@@ -0,0 +1,307 @@
+from typing import Callable, Dict, Tuple, Type
+from pathlib import Path
+
+from junit_helpers import JUnitResultBuilder, JUnitTestCase, JUnitTestSuite, JUnitTestStatus, LOG_FILE_NAME
```
Have we considered using Jenkins plugins instead of custom scripts to format the results?

According to the INFRA page, the Jenkins JUnit Plugin is installed for the ci-cassandra.
https://cwiki.apache.org/confluence/display/INFRA/ci-cassandra.apache.org

This plugin natively supports JUnit XML report files to print the report:
https://plugins.jenkins.io/junit/

Mmuzaf replied:
I see it was used in the `Jenkinsfile` before, but it didn't suit our needs, did it?

```
junit testResults: '**/build/test/**/TEST*.xml,**/cqlshlib.xml,**/nosetests.xml', testDataPublishers: [[$class: 'StabilityTestDataPublisher']]
```

michaelsembwever replied:
`junit testResults` is used in the Jenkinsfile to provide the "Test Result" UI.
e.g. https://ci-cassandra.apache.org/view/patches/job/Cassandra-5-devbranch/92/testReport/

while this collects our many thousands of xml unit test files, it provides no way of aggregating them into one, nor of exporting the results (either whole or per test-type) in aggregated xml unit file. 

the structure and content of the ci_summary.html and results_details tarball is becoming standardised across this and apple's CI system.  it requires aggregated unit results (single xml and folder of html) per test type.

the jenkins UI is not always available.  i.e. we could be running this Jenkinsfile in a private jenkins instance, and the ci_summary.html and results_details tarball are the only two files we have to then share.  in [CASSANDRA-18145](https://issues.apache.org/jira/browse/CASSANDRA-18145) we'll provide `.build/run-ci.sh …` script that setups up in k8s a fresh jenkins, runs the pipeline, gets these two files, tears down the jenkins for you, so that you never have to open the jenkins UI at all (everything on the command line).

also note, aggregating all our test results together requires >14GB ram (all plugins use xerces which isn't streaming). 

a lot of these requirements and challenges are written up in [CASSANDRA-18137](https://issues.apache.org/jira/browse/CASSANDRA-18137)

Mmuzaf commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +12,529 @@
+}
+
+def copyToNightlies(sourceFiles, remoteDirectory='') {
+    if (isCanonical() && sourceFiles?.trim()) {
```
Indentation :-)

Mmuzaf reviewed (approved):
+1

HadesArchitect commented on .build/docker/_docker_run.sh:
```diff
@@ -26,10 +26,14 @@
 # variables, with defaults
 [ "x${cassandra_dir}" != "x" ] || cassandra_dir="$(readlink -f $(dirname "$0")/../..)"
 [ "x${build_dir}" != "x" ] || build_dir="${cassandra_dir}/build"
+[ "x${m2_dir}" != "x" ] || m2_dir="${HOME}/.m2/repository"
```
Does it work on Jenkins? HOME directory wasn't writable for agent, that's why I moved it

michaelsembwever replied:
yes, because jenkins defines it here: https://github.com/apache/cassandra/pull/2852/files#diff-917491029f94ac1ce6c114a542d2a8738ce1f2d1e866eaffe7189d2139250200R322
and here: https://github.com/apache/cassandra/pull/2852/files#diff-917491029f94ac1ce6c114a542d2a8738ce1f2d1e866eaffe7189d2139250200R362

HadesArchitect replied:
Thank you

HadesArchitect commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +12,531 @@
+//  For example 'arm64' requires the labels: cassandra-arm64-small, cassandra-arm64-medium, cassandra-arm64-large.
+//
+// Plugins required are:
+//  git, workflow-job, workflow-cps, junit, workflow-aggregator, ws-cleanup, pipeline-build-step, test-stability, copyartifact.
```
Do we have to link cassius here?

michaelsembwever replied:
that happens in [CASSANDRA-18145](https://issues.apache.org/jira/browse/CASSANDRA-18145)

HadesArchitect commented on .jenkins/Jenkinsfile:
```diff
@@ -11,762 +12,531 @@
+    string(name: 'profile_custom_regexp', defaultValue: '', description: 'Regexp for stages when using custom profile. See `testSteps` in Jenkinsfile for list of stages. Example: stress.*|jvm-dtest.*')
+
+    choice(name: 'architecture', choices: archsSupported() + "all", description: 'Pick architecture. The ARM64 is disabled by default at the moment.')
+    string(name: 'jdk', defaultValue: "", description: 'Restrict JDK versions. (e.g. "11", "17", etc)')
```
Not clear how to pick multiple archs (before there were "all" option in dropdown)

michaelsembwever replied:
architecture still has the "all" option.

jdk is default "all" (if left blank).  but you can now restrict it to a specific jdk.

HadesArchitect replied:
I'd mention it in the description of the field
